src/test/unit/flight_failsafe_unittest.cc

```diff
@@ -256,16 +267,146 @@ TEST(FlightFailsafeTest, TestFailsafeCausesLanding)
+
+        failsafeOnRxCycleStarted();
+        // no call to failsafeOnValidDataReceived();
+
+        failsafeUpdateState();
+
+        // then
+        EXPECT_EQ(FAILSAFE_RX_LOSS_MONITORING, failsafePhase());
+        EXPECT_EQ(true, failsafeIsActive());
+        EXPECT_EQ(1, CALL_COUNTER(COUNTER_MW_DISARM));
+        EXPECT_TRUE(ARMING_FLAG(PREVENT_ARMING));
+    }
+
+    // when
+    for (int i = 0; i < callsToMakeToSimulateTwentyFiveSeconds - 1; i++) {
+
+        failsafeOnRxCycleStarted();
+        failsafeOnValidDataReceived();
+
+        failsafeUpdateState();
+    }
+
+    // then
+    EXPECT_EQ(false, failsafeIsActive());
+    EXPECT_EQ(FAILSAFE_IDLE, failsafePhase());
+    EXPECT_EQ(1, CALL_COUNTER(COUNTER_MW_DISARM)); // disarm not called repeatedly.
+    EXPECT_FALSE(ARMING_FLAG(PREVENT_ARMING));
+}
+
+TEST(FlightFailsafeTest, TestFailsafeDetectsRxLossAndKillswitchEvent)
+{
+    // given
+    int callsToMakeToSimulateTwoSeconds = FAILSAFE_UPDATE_HZ * 2;
+    ENABLE_ARMING_FLAG(ARMED);
+    resetCallCounters();
+
+    // and
+    failsafeStartMonitoring();
+    testMinThrottle = 1100;
+    rcCommand[THROTTLE] = testMinThrottle - 1;  // just below configured min throttle to go for a failsafe just-disarm procedure
+    failsafeConfig.failsafe_kill_switch = 1;    // configure AUX switch as kill switch
+    ACTIVATE_RC_MODE(BOXFAILSAFE);              // and activate it
+
+    //
+    // currently 20 cycles must occur before the lack of an update triggers RX loss detection.
+    //
+    // FIXME see comments about RX_SERIAL/RX_MSP above, the test should likely deal with time rather than counters.
+    int failsafeCounterThreshold = 20;
+
+    // when
+    for (int i = 0; i < failsafeCounterThreshold; i++) {
+
+        failsafeOnRxCycleStarted();
+        // no call to failsafeOnValidDataReceived();
+
+        failsafeUpdateState();
+
+        // then
+        EXPECT_EQ(FAILSAFE_IDLE, failsafePhase());
+        EXPECT_EQ(false, failsafeIsActive());
+    }
+
+    // when
+    for (int i = 0; i < FAILSAFE_UPDATE_HZ - failsafeCounterThreshold; i++) {
+
+        failsafeOnRxCycleStarted();
+        // no call to failsafeOnValidDataReceived();
+
+        failsafeUpdateState();
+
+        // then
+        EXPECT_EQ(FAILSAFE_RX_LOSS_MONITORING, failsafePhase());
+        EXPECT_EQ(true, failsafeIsActive());
+        EXPECT_EQ(1, CALL_COUNTER(COUNTER_MW_DISARM));
+        EXPECT_TRUE(ARMING_FLAG(PREVENT_ARMING));
+    }
+
+    // when
+    rcModeActivationMask = 0;                   // de-activate BOXFAILSAFE
+    for (int i = 0; i < callsToMakeToSimulateTwoSeconds - 1; i++) {
+
+        failsafeOnRxCycleStarted();
+        failsafeOnValidDataReceived();
+
+        failsafeUpdateState();
+    }
+
+    // then
+    EXPECT_EQ(false, failsafeIsActive());
+    EXPECT_EQ(FAILSAFE_IDLE, failsafePhase());
+    EXPECT_EQ(1, CALL_COUNTER(COUNTER_MW_DISARM)); // disarm not called repeatedly.
+    EXPECT_FALSE(ARMING_FLAG(PREVENT_ARMING));
 }
```

Comment:
counters everywhere!  yes.  the failsafe needs to move away from counters and release itself from expected call frequency problem it has.